src/Compilers/CSharp/Test/Semantic/Semantics/StackAllocInitializerTests.cs

```diff
@@ -62,6 +62,9 @@ unsafe void M()
                 // (10,29): error CS0208: Cannot take the address of, get the size of, or declare a pointer to a managed type ('RefG<string>')
                 //         var x2 = stackalloc RefG<string>[10];
                 Diagnostic(ErrorCode.ERR_ManagedAddr, "RefG<string>").WithArguments("RefG<string>").WithLocation(10, 29),
+                // (12,29): error CS0208: Cannot take the address of, get the size of, or declare a pointer to a managed type ('TypedReference')
```

Comment:
>// (12,29): error CS0208: Cannot take the address of, get the size of, or declare a pointer to a managed type ('TypedReference') [](http://example.com/codeflow?start=16&length=128)

It looks like this is a "hard" break for a scenario that used to compile. Not sure if it would work at runtime though.

Reply:
stackalloc wouldn't work at runtime today but could be supported at some point (from offline discussion with David W as part of PR [https://github.com/dotnet/roslyn/pull/64294](https://github.com/dotnet/roslyn/pull/64294)). That's why we kept the `CheckManagedAddr` check an error for stackalloc rather than downgrading it to a warning.

Reply:
Looking at the implementation of TypedReference in the runtime repo, the type doesn't actually contain a reference yet (it's IntPtr). So the error David described wouldn't kick in yet. I'll add stackalloc to the break doc.

Reply:
I would think `stackalloc` should not work for any of the special types `TypedReference`, `ArgIterator` and `RuntimeArgumentHandle` as these can all contain references that only the runtime may understand (and thus keep correct with objects moving around). These types would have to be on the stack as their appropriate types for the runtime to be able to understand them, meaning that they shouldn't be able to be stackalloc'd as it is always an invalid operation currently (future .NET core may allow this, but this probably would not be emitted in its current form since it doesn't provide information to the runtime about what type it's allocating on the stack, and thus has no idea what references it must keep track of). The checks for these types should always disallow these 3 types specifically since current and older runtimes have the fields on these types as IntPtrs or void*s, but with the implementation detail that they may actually be updated by the runtime at any point, meaning they are managed fields (if they never change in practice, that is an implementation detail). Note that in the ECMA-335 spec these types are called byref-like types (I.8.2.1.1), so any restrictions that applies to spans (which are post ECMA-335) should also apply to these types (note also, that the ECMA-335 addendum doesn't change this).

Why did the runtime team want them (`ArgIterator` and `RuntimeArgumentHandle`) to remain unmanaged for now, this is doesn't seem right to me based on that they are the longest-existing byref-like types of all. I also do not see what functional difference it would make other than disallowing always incorrect behaviour from being legal in C# without any warning or error.

Reply:
Tagging @AaronRobinsonMSFT in case anything to add.
I'll quote from our chat, "For the current implementation and supported platforms they [ArgIterator and RuntimeArgumentHandle] can be considered unmanaged without issue - they will always point to the stack. I do not expect that we would ever run into a problem with unmanaged constraint for these two."

Reply:
I see, that's fair then. I think there should be an addition to the [ECMA-335 addendum](https://github.com/dotnet/runtime/blob/main/docs/design/specs/Ecma-335-Augments.md) that these types do not contain any tracking information so that this is actually documented and guaranteed.

Reply:
This is ill-advised. The ECMA-335 is designed to be agnostic to OS implementation details - even though we violate that occasionally. At present the entire varargs scenario is only supported on Windows and is primarily for the C++/CLI scenario. We could update things and add notes but unless people are using .NET Core and actively innovating in the C++/CLI domain, I don't think doing much them at present is appropriate. The only reason this even came up was because of `TypedReference` improvements and @jcouv being curious :) Once we officially decide the varargs message going forward, I agree we should update all of this.

Reply:
> Tagging @AaronRobinsonMSFT in case anything to add. I'll quote from our chat, "For the current implementation and supported platforms they [ArgIterator and RuntimeArgumentHandle] can be considered unmanaged without issue - they will always point to the stack. I do not expect that we would ever run into a problem with unmanaged constraint for these two."

@jcouv Just thought I'd mention that this may not be the case with async2.

/cc @agocke

Reply:
As far as async2 and `ArgIterator` and `RuntimeArgumentHandle`, it is unlikely and I would push back on adding support for use of either of these types in those scenarios.